routes/user.js

```diff
@@ -6,95 +6,63 @@ const crypto = require("crypto");
 const EmailUtils = require("../email-utils");
-const Subscribers = require("../subscribers");
 
-// We send verification emails to addresses that want to subscribe.
-// These addresses are temporarily stored here, mapped to the unique
-// string token that is used for verification.
-const gUnverifiedEmails = new Map();
+const ResponseCodes = Object.freeze({
+  InternalError: 999,
+  EmailNotProvided: 100,
+  EmailNotFound: 101,
+  TokenMismatch: 102,
+});
+
+router.post("/add", async (req, res) => {
+  const user = await models.User.create({ email: req.body.email });
+  const url = `${AppConstants.SERVER_URL}/user/verify?state=${user.verificationToken}&email=${user.email}`;
 
-router.post("/add", (req, res) => {
-  // TODO: use a hash of the email address instead of a random string.
-  const state = crypto.randomBytes(40).toString("hex");
-  req.session.state = state;
-  const email = req.body.email;
-  req.session.email = email;
-  const url = `${AppConstants.SERVER_URL}/user/verify?state=${state}`;
+  try {
+    await EmailUtils.sendEmail(user.email, "Firefox Breach Alert",
+      `Visit this link to subscribe: ${url}`);
 
-  EmailUtils.sendEmail(email, "Firefox Breach Alert",
-    `Visit this link to subscribe: ${url}`)
-    .then(info => {
-      // TODO: set a timer to clear this after an arbitrary timeout period.
-      gUnverifiedEmails.set(state, email);
-      res.json({
-        email,
-        info: "sent verification link",
-        // Send the would-be link back to the client in dummy mode.
-        // eslint-disable-next-line no-process-env
-        link: process.env.DEBUG_DUMMY_SMTP ? url : undefined,
-      });
-    })
-    .catch(error => {
-      res.json({ email, info: "failed to send verification link", error });
-      throw error;
+    res.status(202).json({
+      info: "Sent verification link",
+      // Send the would-be link back to the client in dummy mode.
+      // eslint-disable-next-line no-process-env
+      link: process.env.DEBUG_DUMMY_SMTP ? url : undefined,
+    });
+  } catch (e) {
+    console.log(e);
+    res.status(500).json({
+      error_code: ResponseCodes.InternalError,
+      info: "SMTP error.",
     });
+  }
 });
 
-router.get("/verify", (req, res) => {
-  const email = gUnverifiedEmails.get(req.query.state);
-  if (email) {
-    gUnverifiedEmails.delete(req.query.state);
-    Subscribers.addUser(email).then(() => {
-      res.json({ email, info: `Successfully added ${email}`});
-    }).catch((error) => {
-      res.json({ info: "Failed to add email to database.", error });
+router.get("/verify", async (req, res) => {
+  // eslint-disable-next-line prefer-const
+  const user = await models.User.findOne({ where: { email: req.query.email, verificationToken: req.query.state } });
+  if (user === null) {
+    res.status(400).json({
+      error_code: ResponseCodes.EmailNotFound,
+      info: "Email not found or verification token does not match.",
     });
     return;
+  } else {
```

Comment:
nit: Probably don't need this `else` line, since the `if` statement above already `return`ed.